src/librustc_trans/trans/base.rs

```diff
@@ -2423,12 +2423,14 @@ fn register_fn<'a, 'tcx>(ccx: &CrateContext<'a, 'tcx>,
                          node_id: ast::NodeId,
                          node_type: Ty<'tcx>)
                          -> ValueRef {
-    match node_type.sty {
-        ty::ty_bare_fn(_, ref f) => {
-            assert!(f.abi == Rust || f.abi == RustCall);
+    if let ty::ty_bare_fn(_, ref f) = node_type.sty {
+        if f.abi != Rust && f.abi != RustCall {
+            ccx.sess().span_fatal(sp, "this function may only have \"Rust\" or \"rust-call\" \
```

Comment:
This should be a `span_bug`, not `span_fatal`, unless you think this is user error (the assert suggests it's a bug in the compiler).

Also, it's always nice to print out the value of `f.abi` too.